cockatrice/src/filter_string.cpp

```diff
@@ -0,0 +1,216 @@
+#include "filter_string.h"
+#include "../../common/lib/peglib.h"
+
+#include <QByteArray>
+#include <QString>
+#include <cmath>
+#include <functional>
+
+peg::parser search(R"(
+Start <- QueryPartList
+~ws <- [ ]+
+QueryPartList <- ComplexQueryPart ( ws (ws "and ws")? ComplexQueryPart)* ws*
+
+ComplexQueryPart <- SomewhatComplexQueryPart ws $or<[oO][rR]> ws SomewhatComplexQueryPart / SomewhatComplexQueryPart
+
+SomewhatComplexQueryPart <- [(] QueryPartList [)] / QueryPart
+
+QueryPart <- NotQuery / SetQuery / CMCQuery / PowerQuery / ToughnessQuery / ColorQuery / TypeQuery / OracleQuery / FieldQuery / GenericQuery
+
+NotQuery <- ('not' ws/'-') QueryPart
+SetQuery <- 'e' [:] FlexStringValue
+OracleQuery <- 'o' [:] RegexString
+
+
+CMCQuery <- 'cmc' ws? NumericExpression
+PowerQuery <- 'pow' ws? NumericExpression
+ToughnessQuery <- 'tou' ws? NumericExpression
+
+
+TypeQuery <- [tT] 'ype'? [:] StringValue
+
+Color <- < [Ww] 'hite'? / [Uu] / [Bb] 'lack'? / [Rr] 'ed'? / [Gg] 'reen'?  / [Bb] 'lue'? >
+ColorEx <- Color / [mc]
+
+ColorQuery <- [cC] 'olor'? [:] ColorEx
+
+FieldQuery <- String [:] RegexString / String ws? NumericExpression
+
+NonQuote <- !["].
+UnescapedStringListPart <- [a-zA-Z]+
+String <- UnescapedStringListPart / ["] <NonQuote*> ["]
+StringValue <- String / [(] StringList [)]
+StringList <- StringListString (ws? [,] ws? StringListString)*
+StringListString <- UnescapedStringListPart
+GenericQuery <- RegexString
+RegexString <- String
+
+FlexStringValue <- CompactStringSet / String / [(] StringList [)]
+CompactStringSet <- StringListString ([,+] StringListString)+
+
+NumericExpression <- NumericOperator ws? NumericValue
+NumericOperator <- [=:] / <[><!][=]?>
+NumericValue <- [0-9]+ 
+
+    )");
+
+std::once_flag init;
+
+FilterString::FilterString(const QString &expr)
```

Comment:
Would _really_ like some comments in here

Reply:
Should I add a tip of the day?